src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu

```diff
@@ -67,65 +70,139 @@ namespace cuda
+        using MatrixXcucd = Eigen::Matrix<cuDoubleComplex, Eigen::Dynamic, Eigen::Dynamic>;
+        using Tensor3Xcucd = Eigen::Matrix<VectorXcucd, Eigen::Dynamic, Eigen::Dynamic>;
+
+        int integration_baselines = 0;
+        Tensor3Xcucd integration_data; //TODO: incomplete
+        Eigen::Map<MatrixXcucd> avg_data = Eigen::Map<MatrixXcucd>(pavg_data, avg_dataRows, avg_dataCols);
 
-        // loop over baselines
-        for(int baseline = 0; baseline < integration.baselines; ++baseline)
+        /// loop over baselines
+        for(int baseline = 0; baseline < integration_baselines; ++baseline)
         {
-            // For baseline
-            const double pi = boost::math::constants::pi<double>();
-            double shiftFactor = -2 * pi * uvw[baseline].get()[2] - metadata.oldUVW[baseline].get()[2]; // check these are correct
-            shiftFactor = shiftFactor + 2 * pi * (metadata.GetConstants().phase_centre_ra_rad * metadata.oldUVW[baseline].get()[0]);
-            shiftFactor = shiftFactor -2 * pi * (direction.get()[0] * uvw[baseline].get()[0] - direction.get()[1] * uvw[baseline].get()[1]);
+            const double pi = CUDART_PI;
+            double shiftFactor = -2 * pi * uvw[baseline].z - oldUVW[baseline].z;
+            shiftFactor = shiftFactor + 2 * pi * (constants.phase_centre_ra_rad * oldUVW[baseline].x);
+            shiftFactor = shiftFactor - 2 * pi * (direction.x * uvw[baseline].x - direction.y * uvw[baseline].y);
 
-            if(baseline % 1000 == 1)
+            // loop over channels
+            for(int channel = 0; channel < constants.channels; channel++)
             {
-                std::cout << "ShiftFactor for baseline " << baseline << " is " << shiftFactor << std::endl;
-            }
+                double shiftRad = shiftFactor / constants.GetChannelWavelength(channel);
 
-            // Loop over channels
-            for(int channel = 0; channel < metadata.GetConstants().channels; channel++)
-            {
-                double shiftRad = shiftFactor / metadata.GetConstants().channel_wavelength[channel];
-                std::complex<double> v = data(channel, baseline);
+                cuDoubleComplex exp = cuCexp(make_cuDoubleComplex(0.0, shiftRad));
+                for(int i = 0; i < integration_data(channel, baseline).cols(); i++)
+                {
+                    integration_data(channel, baseline)(i) = cuCmul(integration_data(channel, baseline)(i), exp);
+                }
 
-                data(channel, baseline) = v * std::exp(std::complex<double>(0.0, 1.0) * std::complex<double>(shiftRad, 0.0));
-                if(data(channel, baseline).real() == NAN
-                || data(channel, baseline).imag() == NAN)
+                for(int i = 0; i < integration_data(channel, baseline).cols(); i++)
                 {
-                    metadata.avg_data(1, baseline) += data(channel,baseline);
+                    //if(!integration_data(channel, baseline).hasNaN())
```

Comment:
the cuDoubleComplex type seems to be lacking a lot of operator overrides. hasNaN() generates this error and I didnt bother fixing it as I needed to swap to tensors:
```
error: no operator "==" matches these operands
            operand types are: const cuDoubleComplex == const cuDoubleComplex
          detected during:
...
instantiation of "__nv_bool Eigen::DenseBase<Derived>::hasNaN() const [with Derived=Eigen::Matrix<cuDoubleComplex, -1, 1, 0, -1, 1>]"
```

Reply:
basically the kernel compiler does not support std::complex<T> and looking online its recommended to use `cuFloatComplex` and `cuDoubleComplex` whose header looks like this: http://pixel.ecn.purdue.edu:8080/purpl/WSJ/projects/DirectionalStippling/include/cuComplex.h 

It might be possible to write C++ style extension operator overrides for this so I'll give it go in a later branch

Reply:
OK, so this will be eventually tackled, right? Maybe add a TODO to make it explicit that you mean to add the condition.